File: apps/driver-android/src/main/kotlin/lk/mageride/driver/ride/ActiveRideScreen.kt
package lk.mageride.driver.ride

import android.content.Intent
import android.net.Uri
import androidx.annotation.StringRes
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Call
import androidx.compose.material.icons.outlined.Navigation
import androidx.compose.material.icons.outlined.Person
import androidx.compose.material.icons.outlined.Shield
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.R
import lk.mageride.driver.home.DriverHomeMap
import lk.mageride.driver.ui.MoneyFormat
import lk.mageride.driver.ui.component.DashboardBanner
import lk.mageride.driver.ui.component.DashboardSheet
import lk.mageride.driver.ui.component.MageRideCta
import lk.mageride.driver.ui.component.StatusCta
import lk.mageride.driver.ui.theme.ControlTokens
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.RideState
import lk.mageride.shared.data.models.Ulid
import lk.mageride.shared.data.models.ride.RideKind
import org.koin.androidx.compose.koinViewModel
import org.koin.core.parameter.parametersOf

/**
 * **SCR-DA-015 · the active ride** — `Accepted → DriverArrived → InProgress → Completed`.
 *
 * The wireframe: a navigation banner, the map with the pickup pin, and a sheet carrying the rider,
 * the route and the fare, a **Call rider** / **SOS** pair, and one big CTA whose label is the ride's
 * own next step.
 *
 * **The QR confirm sheet is not an extra** (AL-47, US-26.1). A ride the passenger paid by scanning
 * the driver's own bank LankaQR has no gateway callback that could settle it — the money moved
 * bank-to-bank and only the driver's banking app saw it — so *"QR payment received?"* **is** the
 * settlement, and confirming is what posts the earning (R-05).
 *
 * @param onFinished The ride reached a terminal state; the driver goes back to standby.
 */
@Composable
internal fun ActiveRideScreen(rideId: Ulid, onFinished: () -> Unit, modifier: Modifier = Modifier) {
    val viewModel: ActiveRideViewModel = koinViewModel { parametersOf(rideId) }
    val state by viewModel.state.collectAsStateWithLifecycle()
    val context = LocalContext.current

    Column(modifier = modifier.fillMaxSize()) {
        DashboardBanner(
            text = stringResource(navigationHint(state.rideState)),
            accent = MaterialTheme.colorScheme.secondary,
            icon = Icons.Outlined.Navigation,
        )

        state.error?.let { message ->
            DashboardBanner(text = stringResource(message), accent = MaterialTheme.colorScheme.error)
        }

        Box(modifier = Modifier.weight(1f)) {
            // MAP-10's 100 m circle, drawn where the driver is being sent: the pickup until they
            // are on board, the drop afterwards. It is D5' §7's arrival geofence — the same circle
            // that decides `DriverArrived` — so the driver can see what they have to reach.
            DriverHomeMap(
                position = state.position,
                vehicleType = null,
                geofence = state.geofence,
                modifier = Modifier.fillMaxSize(),
            )
        }

        ActiveRideSheet(state = state, viewModel = viewModel)
    }

    RideDialogs(state = state, viewModel = viewModel)

    // AL-48's direct dial. The number is the counterparty's real MSISDN, revealed post-accept, and
    // on a proxy booking it is the RIDER's rather than the booker's (P-05) — ride-svc decides that,
    // not this screen.
    LaunchedEffect(state.dialNumber) {
        state.dialNumber?.let { number ->
            runCatching {
                context.startActivity(
                    Intent(Intent.ACTION_DIAL, Uri.parse("tel:$number")).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK),
                )
            }
            viewModel.consume()
        }
    }

    LaunchedEffect(state.finished) {
        if (state.finished) onFinished()
    }
}

/** The wireframe's sheet: rider, route, the two icon buttons, and the state's own CTA. */
@Composable
private fun ActiveRideSheet(state: ActiveRideState, viewModel: ActiveRideViewModel, modifier: Modifier = Modifier) {
    val ride = state.ride

    DashboardSheet(modifier = modifier) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            Surface(
                modifier = Modifier.size(ControlTokens.AvatarSmall),
                shape = CircleShape,
                color = MaterialTheme.colorScheme.surfaceVariant,
            ) {
                Icon(
                    imageVector = Icons.Outlined.Person,
                    contentDescription = null,
                    modifier = Modifier.size(ControlTokens.RowIcon),
                    tint = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = ride?.riderName ?: stringResource(R.string.ride_rider),
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSurface,
                )
                Text(
                    text = listOfNotNull(
                        ride?.pickup?.address,
                        ride?.dropoff?.address,
                    ).joinToString(RideLabels.ROUTE_ARROW) +
                        (ride?.fare?.amountMinor?.let { " · ${MoneyFormat.rupees(it)}" }.orEmpty()),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            OutlinedButton(
                onClick = { viewModel.open(RideSheet.CALL_TYPE) },
                modifier = Modifier.weight(1f),
                enabled = ride?.counterpartyPhone != null,
            ) {
                Icon(imageVector = Icons.Outlined.Call, contentDescription = null)
                Text(
                    text = stringResource(
                        if (ride?.kind == RideKind.PACKAGE) R.string.ride_call_sender else R.string.ride_call,
                    ),
                    modifier = Modifier.padding(start = MageRideTheme.spacing.xxs),
                )
            }
            OutlinedButton(
                onClick = viewModel::triggerSos,
                modifier = Modifier.weight(1f),
                enabled = state.position != null,
            ) {
                Icon(
                    imageVector = Icons.Outlined.Shield,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.error,
                )
                Text(
                    text = stringResource(R.string.ride_sos),
                    color = MaterialTheme.colorScheme.error,
                    modifier = Modifier.padding(start = MageRideTheme.spacing.xxs),
                )
            }
        }

        // D2' §SCR-DA-015's table colours the two ends of the ride by consequence: Start is
        // `success`, End is `error`. Arriving is neither — it is the ordinary primary CTA.
        val cta = forwardAction(state)
        when {
            cta != null && state.rideState == RideState.Accepted -> MageRideCta(
                label = stringResource(cta),
                onClick = viewModel::advance,
                loading = state.busy,
            )

            cta != null -> StatusCta(
                label = stringResource(cta),
                accent = if (state.rideState == RideState.InProgress) {
                    MaterialTheme.colorScheme.error
                } else {
                    MageRideTheme.status.success
                },
                onClick = viewModel::advance,
                enabled = !state.busy,
            )

            state.awaitingQrConfirm -> MageRideCta(
                label = stringResource(R.string.ride_qr_title),
                onClick = { viewModel.open(RideSheet.QR_CONFIRM) },
                loading = state.busy,
            )
        }
    }
}

/**
 * The CTA's label, from the ride's own state.
 *
 * `null` means there is no forward move from here — a terminal ride, or a driver-QR ride waiting on
 * AL-47's confirm, which is a different button.
 */
@StringRes
private fun forwardAction(state: ActiveRideState): Int? = when (state.rideState) {
    RideState.Accepted -> R.string.ride_arrived
    RideState.DriverArrived -> R.string.ride_start
    RideState.InProgress -> R.string.ride_complete
    else -> null
}

/** The banner above the map — where the driver is being sent next. */
@StringRes
private fun navigationHint(state: RideState?): Int = when (state) {
    RideState.Accepted, RideState.DriverArrived -> R.string.ride_navigate_pickup
    RideState.InProgress -> R.string.ride_navigate_drop
    else -> R.string.ride_finished
}

/** Symbols rather than sentences — the same rule `Rs` and `L3` follow (C068). */
internal object RideLabels {

    /** The wireframe's `Galle Face → Nugegoda`. An arrow reads the same in Si, Ta and En. */
    const val ROUTE_ARROW: String = " → "
}

File: apps/driver-ios/DriverApp/Ride/ActiveRideScreen.swift
import MageRideShared
import SwiftUI

/// **SCR-DI-015 · the active ride** — `Accepted → DriverArrived → InProgress → Completed`.
///
/// The wireframe: a navigation banner, the map with the pickup pin, and a sheet carrying the rider, the
/// route and the fare, a **Call rider** / **SOS** pair, and one big CTA whose label is the ride's own
/// next step.
///
/// **The QR confirm sheet is not an extra** (AL-47, US-26.1). A ride the passenger paid by scanning the
/// driver's own bank LankaQR has no gateway callback that could settle it — the money moved
/// bank-to-bank and only the driver's banking app saw it — so *"QR payment received?"* **is** the
/// settlement, and confirming is what posts the earning (R-05).
///
/// - Parameters:
///   - onFinished: The ride reached a terminal state; the driver goes back to standby.
///   - onOpenVoipCall: SCR-DI-031 (C093) — AL-48's chooser answered **Free call**.
///   - onOpenSos: SCR-DI-032 (C093) — the alarm is a screen with a confirmation, not a button that
///     fires.
@MainActor
struct ActiveRideScreen: View {

    @StateObject private var model: ActiveRideModel

    private let rideId: String
    private let onFinished: () -> Void
    private let onOpenVoipCall: (String) -> Void
    private let onOpenSos: (String) -> Void

    init(
        rideId: String,
        model: @autoclosure @escaping () -> ActiveRideModel,
        onFinished: @escaping () -> Void,
        onOpenVoipCall: @escaping (String) -> Void,
        onOpenSos: @escaping (String) -> Void
    ) {
        self.rideId = rideId
        _model = StateObject(wrappedValue: model())
        self.onFinished = onFinished
        self.onOpenVoipCall = onOpenVoipCall
        self.onOpenSos = onOpenSos
    }

    var body: some View {
        content
            .navigationBarTitleDisplayMode(.inline)
            .task {
                model.start()
                await model.refresh()
            }
            .onDisappear(perform: model.stop)
            .sheet(item: Binding(get: { model.state.sheet }, set: { model.open($0) })) { sheet in
                RideSheets(sheet: sheet, model: model)
            }
            // The two screens C093 owns. Both are takeovers about this ride, so both carry its id.
            .onChange(of: model.state.isVoipRequested) { requested in
                guard requested else { return }
                model.consume()
                onOpenVoipCall(rideId)
            }
            .onChange(of: model.state.isSosRequested) { requested in
                guard requested else { return }
                model.consume()
                onOpenSos(rideId)
            }
            .onChange(of: model.state.isFinished) { finished in
                if finished { onFinished() }
            }
    }

    /// A parcel is not a passenger, and **this screen does not yet know what to do about it.**
    ///
    /// ``DriverRoute/activeRide(rideId:)`` is the destination for every live job — R-01 keeps one
    /// aggregate and ``PushRouter`` resolves `mageride://package/{id}` to it — so the kind is not known
    /// until the ride has been read. On Android, `ActiveRideScreen` reads it and hands over to
    /// `DeliveryScreen` on `RideKind.PACKAGE`; **SCR-DI-016a/b/c is C089's** and there is nothing to
    /// hand over to yet.
    ///
    /// Until it lands a package draws this screen, and the model is already honest about it:
    /// ``ActiveRideState/isPackage`` switches the Call button to *"Call sender"* and switches the poll
    /// **off** (``ActiveRideState/isPollable``), so C089's own loop will be the only one folding server
    /// states onto the ride. The hand-over is one `if` at the top of this property — the same shape the
    /// Android twin has — and it is C089's line to add.
    @ViewBuilder
    private var content: some View {
        VStack(spacing: 0) {
            DashboardBanner(
                text: model.state.navigationHintKey.localised,
                accent: MageRideColor.secondary,
                symbolName: "location.north.line.fill"
            )

            if let errorKey = model.state.errorKey {
                DashboardBanner(text: errorKey.localised, accent: MageRideColor.error)
                    .onTapGesture(perform: model.consume)
            }

            // MAP-10's 100 m circle, drawn where the driver is being sent: the pickup until they are on
            // board, the drop afterwards. It is D5' §7's arrival geofence — the same circle that decides
            // `DriverArrived` — so the driver can see what they have to reach.
            DriverHomeMap(
                position: model.state.position,
                token: nil,
                geofence: model.state.geofence
            )

            sheet
        }
        .background(MageRideColor.background)
    }

    /// The wireframe's sheet: rider, route, the two buttons, and the state's own CTA.
    private var sheet: some View {
        DashboardSheet {
            HStack(spacing: MageRideSpacing.xs) {
                Image(systemName: "person.crop.circle.fill")
                    .font(.system(size: MageRideControl.avatarSmall))
                    .foregroundStyle(MageRideColor.onSurfaceVariant)

                VStack(alignment: .leading, spacing: 1) {
                    Text(model.state.ride?.riderName ?? "ride_rider".localised)
                        .mageFont(.title)
                        .foregroundStyle(MageRideColor.onSurface)
                    Text(model.state.routeLine)
                        .mageFont(.bodySmall)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                }
                Spacer(minLength: 0)
            }
            .accessibilityElement(children: .combine)

            HStack(spacing: MageRideSpacing.xs) {
                outlinedAction(
                    labelKey: model.state.isPackage ? "ride_call_sender" : "ride_call",
                    symbolName: "phone.fill",
                    tint: MageRideColor.onSurface,
                    isEnabled: model.state.ride?.counterpartyPhone != nil
                ) {
                    model.open(.callType)
                }

                outlinedAction(
                    labelKey: "ride_sos",
                    symbolName: "shield.lefthalf.filled",
                    tint: MageRideColor.error,
                    // SCR-DI-032 needs a fix to attach to the alarm — `POST /v1/sos` has no
                    // positionless form.
                    isEnabled: model.state.position != nil,
                    action: model.openSos
                )
            }

            forwardCta
        }
    }

    /// D2' §SCR-DI-015's table colours the two ends of the ride by consequence: Start is `success`,
    /// Complete is `error`. Arriving is neither — it is the ordinary primary CTA.
    @ViewBuilder
    private var forwardCta: some View {
        if let actionKey = model.state.forwardActionKey {
            Button { Task { await model.advance() } } label: {
                Text(key: actionKey)
            }
            .buttonStyle(ctaStyle)
            .disabled(model.state.isBusy)
        } else if model.state.awaitingQrConfirm {
            Button { model.open(.qrConfirm) } label: {
                Text(key: "ride_qr_title")
            }
            .buttonStyle(.mageCta(loading: model.state.isBusy))
            .disabled(model.state.isBusy)
        }
    }

    private var ctaStyle: MageRideCtaStyle {
        guard let rideState = model.state.rideState else { return .mageCta(loading: model.state.isBusy) }
        switch rideState {
        case RideState.driverArrived: return .mageCtaStatus(MageRideColor.success, loading: model.state.isBusy)
        case RideState.inProgress: return .mageCtaStatus(MageRideColor.error, loading: model.state.isBusy)
        default: return .mageCta(loading: model.state.isBusy)
        }
    }

    /// The wireframe's `btn-out` — an outlined half-width action.
    private func outlinedAction(
        labelKey: String,
        symbolName: String,
        tint: Color,
        isEnabled: Bool,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            HStack(spacing: MageRideSpacing.xxs) {
                Image(systemName: symbolName)
                    .font(.footnote)
                Text(key: labelKey)
                    .mageFont(.bodySmall)
            }
            .foregroundStyle(tint)
            .frame(maxWidth: .infinity, minHeight: MageRideControl.minimumTapTarget)
            .overlay {
                RoundedRectangle(cornerRadius: MageRideRadius.sm, style: .continuous)
                    .strokeBorder(MageRideColor.outline, lineWidth: 1)
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(!isEnabled)
        .opacity(isEnabled ? 1 : 0.4)
    }
}
